### PR TITLE
MGMT-17120: Delete device's associated EnrollmentRequest

### PR DESCRIPTION
**Pull request by oglok**
If Device resource is deleted, and its associated EnrollmentRequest persists, a re-enrollment process of the device won't be possible. This makes sure that enrollment requests are deleted with its device.

### CODE REVIEW

File: internal/service/device.go [hunk 1]
```diff
@@ -122,6 +122,11 @@ func (h *ServiceHandler) DeleteDevice(ctx context.Context, request server.Delete
 	err := h.store.Device().Delete(ctx, orgId, request.Name, h.taskManager.DeviceUpdatedCallback)
 	switch err {
 	case nil:
+		// If the device deletion was successful, delete the associated EnrollmentRequest
```
Comment (mangelajo) [inline internal/service/device.go:125]: I would try to wrap the device delete and enrollment request delete in a transaction if we can.  Then a device cannot be deleted, something go wrong in the middle and the enrollment request left around.

**Comment by oglok** [comment]
let me know what you think now @mangelajo @hexfusion

**Review by hexfusion — COMMENTED** [review]
Seems reasonable for a cascading like delete defer to @avishayt or @mangelajo for the tag

**Comment by mangelajo** [comment]
> Seems reasonable for a cascading like delete defer to @avishayt or @mangelajo for the tag

What do you mean?

**Comment by hexfusion** [comment]
I just looked at the code briefly but I am not the authority for store. So wanted to make sure you or @avishayt didn't have opinions before it merged

**Comment by oglok** [comment]
I forgot to sign the commit. On my way...

File: internal/store/device.go [hunk 2]
```diff
@@ -247,14 +247,25 @@ func (s *DeviceStore) UpdateStatus(ctx context.Context, orgId uuid.UUID, resourc
...
 		existingRecord = model.Device{Resource: model.Resource{OrgID: orgId, Name: name}}
 		result := innerTx.First(&existingRecord)
 		if result.Error != nil {
 			return result.Error
 		}
-		result = innerTx.Unscoped().Delete(&existingRecord)
-		return result.Error
+
+		associatedRecord = model.EnrollmentRequest{Resource: model.Resource{OrgID: orgId, Name: name}}
+		innerTx.First(&associatedRecord) // Ignore the error if the EnrollmentRequest does not exist
```
[the remark this answers is not in the document]
Comment (oglok) [reply internal/store/device.go]: done!

File: internal/store/device.go [hunk 3]
```diff
@@ -247,14 +247,24 @@ func (s *DeviceStore) UpdateStatus(ctx context.Context, orgId uuid.UUID, resourc
...
+	err := s.db.Transaction(func(innerTx *gorm.DB) error {
 		existingRecord = model.Device{Resource: model.Resource{OrgID: orgId, Name: name}}
 		result := innerTx.First(&existingRecord)
 		if result.Error != nil {
 			return result.Error
 		}
-		result = innerTx.Unscoped().Delete(&existingRecord)
-		return result.Error
+
+		associatedRecord = model.EnrollmentRequest{Resource: model.Resource{OrgID: orgId, Name: name}}
```
Comment (avishayt) [inline internal/store/device.go:259]:
```suggestion
		associatedRecord := model.EnrollmentRequest{Resource: model.Resource{OrgID: orgId, Name: name}}
```

File: internal/store/device.go [hunk 4]
```diff
@@ -247,14 +247,24 @@ func (s *DeviceStore) UpdateStatus(ctx context.Context, orgId uuid.UUID, resourc
 
 func (s *DeviceStore) Delete(ctx context.Context, orgId uuid.UUID, name string, callback DeviceStoreCallback) error {
 	var existingRecord model.Device
-	err := s.db.Transaction(func(innerTx *gorm.DB) (err error) {
+	var associatedRecord model.EnrollmentRequest
+
+	err := s.db.Transaction(func(innerTx *gorm.DB) error {
```
Comment (avishayt) [inline internal/store/device.go:252]: Can revert these lines (see next comment)

File: internal/store/device.go [hunk 5]
```diff
@@ -247,14 +247,24 @@ func (s *DeviceStore) UpdateStatus(ctx context.Context, orgId uuid.UUID, resourc
...
-		result = innerTx.Unscoped().Delete(&existingRecord)
-		return result.Error
+
+		associatedRecord = model.EnrollmentRequest{Resource: model.Resource{OrgID: orgId, Name: name}}
+
+		if err := innerTx.Unscoped().Delete(&existingRecord).Error; err != nil {
+			return err
+		}
+
+		innerTx.Unscoped().Delete(&associatedRecord) // Ignore the error if the EnrollmentRequest does not exist
```
Comment (avishayt) [inline internal/store/device.go:265]:
```suggestion
		err = innerTx.Unscoped().Delete(&associatedRecord) // Ignore the error if the EnrollmentRequest does not exist
		if err != nil {
			// log.Warn?
		}
```

File: internal/store/device.go [hunk 6]
```diff
@@ -247,14 +248,25 @@ func (s *DeviceStore) UpdateStatus(ctx context.Context, orgId uuid.UUID, resourc
 
 func (s *DeviceStore) Delete(ctx context.Context, orgId uuid.UUID, name string, callback DeviceStoreCallback) error {
 	var existingRecord model.Device
+
```
Comment (avishayt) [inline internal/store/device.go:251]:
```suggestion
log := log.WithReqIDFromCtx(ctx, s.log)
```

File: internal/store/device.go [hunk 7]
```diff
@@ -247,14 +248,25 @@ func (s *DeviceStore) UpdateStatus(ctx context.Context, orgId uuid.UUID, resourc
...
-		return result.Error
+
+		associatedRecord := model.EnrollmentRequest{Resource: model.Resource{OrgID: orgId, Name: name}}
+
+		if err := innerTx.Unscoped().Delete(&existingRecord).Error; err != nil {
+			return err
+		}
+
+		if err := innerTx.Unscoped().Delete(&associatedRecord); err != nil {
+			klog.Warningf("Failed to delete associated enrollment request: %v", err)
```
Comment (avishayt) [inline internal/store/device.go:266]:
```suggestion
			log.Warningf("failed to delete associated enrollment request: %v", err)
```

File: internal/store/device.go [hunk 8]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/lib/pq"
 	"github.com/sirupsen/logrus"
 	"gorm.io/gorm"
+	"k8s.io/klog/v2"
```
Comment (avishayt) [inline internal/store/device.go:19]:
```suggestion
```

**Review by avishayt — APPROVED** [review]

**Pull request merged**